zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs

```diff
@@ -0,0 +1,264 @@
+use crate::evm_circuit::execution::ExecutionGadget;
+use crate::evm_circuit::param::{N_BYTES_GAS, N_BYTES_MEMORY_WORD_SIZE};
+use crate::evm_circuit::step::ExecutionState;
+use crate::evm_circuit::util::common_gadget::CommonErrorGadget;
+use crate::evm_circuit::util::constraint_builder::ConstraintBuilder;
+use crate::evm_circuit::util::math_gadget::LtGadget;
+use crate::evm_circuit::util::memory_gadget::{
+    MemoryAddressGadget, MemoryCopierGasGadget, MemoryExpansionGadget,
+};
+use crate::evm_circuit::util::{CachedRegion, Cell, Word};
+use crate::evm_circuit::witness::{Block, Call, ExecStep, Transaction};
+use crate::util::Expr;
+use eth_types::evm_types::{GasCost, OpcodeId};
+use eth_types::{Field, ToLittleEndian};
+use halo2_proofs::circuit::Value;
+use halo2_proofs::plonk::Error;
+
+/// Gadget to implement the corresponding out of gas errors for
+/// [`OpcodeId::CALLDATACOPY`], [`OpcodeId::CODECOPY`] and
+/// [`OpcodeId::RETURNDATACOPY`].
+#[derive(Clone, Debug)]
+pub(crate) struct ErrorOOGMemoryCopyGadget<F> {
+    opcode: Cell<F>,
+    /// Source offset (second stack pop value)
+    src_offset: Word<F>,
+    /// Destination offset and size to copy (first and third stack pop values)
+    dst_memory_addr: MemoryAddressGadget<F>,
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+    memory_copier_gas: MemoryCopierGasGadget<F, { GasCost::COPY }>,
+    insufficient_gas: LtGadget<F, N_BYTES_GAS>,
+    common_error_gadget: CommonErrorGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGMemoryCopyGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasMemoryCopy";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasMemoryCopy;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        cb.require_in_set(
+            "ErrorOutOfGasMemoryCopy opcode must be CALLDATACOPY, CODECOPY or RETURNDATACOPY",
+            opcode.expr(),
+            vec![
+                OpcodeId::CALLDATACOPY.expr(),
+                OpcodeId::CODECOPY.expr(),
+                OpcodeId::RETURNDATACOPY.expr(),
+            ],
+        );
+
+        let dst_offset = cb.query_cell_phase2();
+        let src_offset = cb.query_word_rlc();
+        let copy_size = cb.query_word_rlc();
+
+        cb.stack_pop(dst_offset.expr());
+        cb.stack_pop(src_offset.expr());
+        cb.stack_pop(copy_size.expr());
+
+        let dst_memory_addr = MemoryAddressGadget::construct(cb, dst_offset, copy_size);
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [dst_memory_addr.address()]);
+        let memory_copier_gas = MemoryCopierGasGadget::construct(
+            cb,
+            dst_memory_addr.length(),
+            memory_expansion.gas_cost(),
+        );
+
+        debug_assert!(
```

Comment:
I think just comment that these three opcode has same constant gas is OK.  remove debug_assert.

Reply:
Replaced `debug_assert` with a comment in [99aba46](https://github.com/scroll-tech/zkevm-circuits/pull/368/commits/99aba4648a2423bd565276a03f9d9072c2359690).